### PR TITLE
[Trainer] Add nan/inf logging filter

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Some losses sometimes produce `inf` losses which however doesn't necessarily mean that the training went bad. When using the CTC loss for speech recognition - see: https://github.com/huggingface/transformers/pull/13620 this is often the case. The problem is that as soon as one single loss step is `inf` or `nan` the rest of the training logs will display `inf` or `nan`. In this PR a flag is added that allows the user to filter out `nan` and `inf` values for training. It defaults to `False` and will be set to `True` in all CTC training scripts.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -468,6 +468,15 @@ class TrainingArguments:
     )
     logging_first_step: bool = field(default=False, metadata={"help": "Log the first global_step"})
     logging_steps: int = field(default=500, metadata={"help": "Log every X updates steps."})
+    logging_nan_inf_filter: str = field(
+        default=False,
+        metadata={
+            "help": (
+                "Filter nan and inf losses for logging. "
+                "Note this flag only affects the logging output and not the optimization step."
```
Comment (sgugger) [inline src/transformers/training_args.py:476]: No need to be this long here, but there should be a proper documentation above that can be longer!

**Review by sgugger — APPROVED** [review]
Thanks for adding this! I wonder if the default of the new feature shouldn't be True? It's a little bi breaking, but only for the logging, not the actual training, and it makes more sense to ignore thoses NaN/infs (which are ignored by default by the optimizer since there is some gradient clipping).

**Comment by patrickvonplaten** [comment]
Happy to let it default to True!

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -468,6 +478,7 @@ class TrainingArguments:
     )
     logging_first_step: bool = field(default=False, metadata={"help": "Log the first global_step"})
     logging_steps: int = field(default=500, metadata={"help": "Log every X updates steps."})
+    logging_nan_inf_filter: str = field(default=True, metadata={"help": "Filter nan and inf losses for logging. "})
```
Comment (patrickvonplaten) [inline src/transformers/training_args.py:481]:
```suggestion
    logging_nan_inf_filter: str = field(default=True, metadata={"help": "Filter nan and inf losses for logging."})
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -468,6 +478,7 @@ class TrainingArguments:
     )
     logging_first_step: bool = field(default=False, metadata={"help": "Log the first global_step"})
     logging_steps: int = field(default=500, metadata={"help": "Log every X updates steps."})
+    logging_nan_inf_filter: str = field(default=True, metadata={"help": "Filter nan and inf losses for logging."})
```
Comment (patrickvonplaten) [inline src/transformers/training_args.py:481]: The default is now `True`
